File: app/src/main/java/com/example/github_api_concrete/view/activities/PullRequestsActivity.java
package com.example.github_api_concrete.view.activities;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.example.github_api_concrete.R;
import com.example.github_api_concrete.model.pojo.pullrequests.Response;
import com.example.github_api_concrete.model.pojo.repos.Item;
import com.example.github_api_concrete.view.adapter.PullRequestsRecyclerViewAdapter;
import com.example.github_api_concrete.view.adapter.ReposRecyclerViewAdapter;
import com.example.github_api_concrete.view.interfaces.OnClick;
import com.example.github_api_concrete.view.interfaces.OnClickPR;
import com.example.github_api_concrete.viewmodel.PullRequestsViewModel;
import com.example.github_api_concrete.viewmodel.ReposViewModel;
import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.List;

import static com.example.github_api_concrete.view.activities.ReposActivity.OWNER;
import static com.example.github_api_concrete.view.activities.ReposActivity.REPO;

public class PullRequestsActivity extends AppCompatActivity implements OnClickPR {

    private RecyclerView recyclerViewPR;
    private ProgressBar progressBarPR;
    private PullRequestsViewModel viewModelPR;
    private PullRequestsRecyclerViewAdapter adapterPR;
    private List<Response> pullRequestList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pull_requests);

        initViews();

        Intent intent = getIntent();
        Bundle bundle = intent.getExtras();
        String owner = bundle.getString(OWNER);
        String repo = bundle.getString(REPO);

        recyclerViewPR.setAdapter(adapterPR);
        recyclerViewPR.setLayoutManager(new LinearLayoutManager(this));

        viewModelPR.getAllPR(owner, repo);

        viewModelPR.getListPR().observe(this, pullRequests -> {
            if (pullRequests != null && !pullRequests.isEmpty()){
                adapterPR.updateList(pullRequests);
            } else {
                Snackbar.make(recyclerViewPR, "Não existem pull requests listados neste repositório.", Snackbar.LENGTH_LONG).show();
            }
        });

        viewModelPR.getLoading().observe(this, loading -> {
            if (loading) {
                progressBarPR.setVisibility(View.VISIBLE);
            } else {
                progressBarPR.setVisibility(View.GONE);
            }
        });

    }

    private void initViews() {
        recyclerViewPR = findViewById(R.id.recyclerViewPR);
        progressBarPR = findViewById(R.id.progressBarPR);
        viewModelPR = ViewModelProviders.of(this).get(PullRequestsViewModel.class);
        adapterPR = new PullRequestsRecyclerViewAdapter(pullRequestList, this);
    }

    @Override
    public void click(Response response) {
        String pullRequestPage = response.getHtmlUrl();
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setData(Uri.parse(pullRequestPage));
        startActivity(intent);
    }
}

File: app/src/main/java/com/abs/javarepos/view/PullRequestsActivity.kt
package com.abs.javarepos.view

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.abs.javarepos.R
import com.abs.javarepos.model.PullRequest
import com.abs.javarepos.model.Repo
import com.abs.javarepos.view.adapter.PullRequestAdapter
import com.abs.javarepos.view.util.FadeUtils
import com.abs.javarepos.viewmodel.PullRequestsViewModel
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.activity_pull_requests.*


class PullRequestsActivity : AppCompatActivity() {

    companion object {
        const val REPO_KEY = "repo_key"
    }

    lateinit var repo: Repo
    private lateinit var pullRequestAdapter: PullRequestAdapter
    private lateinit var viewModel: PullRequestsViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pull_requests)

        repo = intent.getSerializableExtra(REPO_KEY) as Repo

        supportActionBar?.title = repo.name

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        pullRequestAdapter = PullRequestAdapter(object : PullRequestAdapter.OnItemClickListener {
            override fun onItemClick(pullRequest: PullRequest) {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(pullRequest.html_url)))
            }
        })

        rvPullRequests.layoutManager = LinearLayoutManager(this)
        rvPullRequests.adapter = pullRequestAdapter

        viewModel = ViewModelProviders.of(this).get(PullRequestsViewModel::class.java)
        viewModel.fetchPullRequests(repo).observe(this, Observer { items ->
            FadeUtils.fadeOut(progressBar)
            pullRequestAdapter.addItems(items)
        })

        viewModel.networkError.observe(this, Observer { hasNetworkError ->
            FadeUtils.fadeOut(progressBar)
            if (hasNetworkError) {
                Snackbar.make(clPullRequests, getString(R.string.error_get_pull_requests), Snackbar.LENGTH_INDEFINITE)
                    .setAction(getString(R.string.action_reload)) {
                        FadeUtils.fadeIn(progressBar)
                        viewModel.fetchPullRequests(repo)
                    }
                    .show()
            }
        })
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }
}
